### ISSUE TITLE
Rendering time on activity pages impacts performance of entire H app

### ISSUE THREAD

**Issue by hmstepanek** [hypothesis/product-backlog#830]
Rendering time for activity pages is to put it simply; out of control. 

Whenever an activity page is accessed that has the max number of annotations (200) it takes > 1s to load. This in turn causes request queueing because none of the other requests can be full-filled until that one is completed. It also eats up the CPU which causes other requests to run slower.

Note in one such [incident](https://insights.newrelic.com/accounts/1385283/dashboards/746090?duration=3600000&end_time=1539100800000&now=1539298216773) the 50th percentile increased by 89%. Requests that typically took 4 ms instead took 39 ms.

Here is an example breakdown of a typical activity page request. Note it took 2.3 s to render the page.
![image](https://user-images.githubusercontent.com/30059933/46838157-dcd66400-cd6c-11e8-92cb-e0bce240809c.png) The max request time today was 4.6s.

The SearchController views are by far our top slowest transactions.
![image](https://user-images.githubusercontent.com/30059933/46838417-0fcd2780-cd6e-11e8-8a87-1a331b10128f.png)

This rendering time also causes UI glitches when a user clicks the back button or x's out a search term and starts typing a new one. It takes so long to load the page that the new search term the user typed into the search box gets deleted when the rendering finally completes.

These long rendering times cause mini-incidents throughout the day that generally aren't reported to on-call staff however, when they become severe enough they do eventually wake someone up. 
![image](https://user-images.githubusercontent.com/30059933/46839148-5cfec880-cd71-11e8-8511-5ceba43abb83.png)

See the following for one such severe incident:
https://hypothes-is.slack.com/archives/C0LUWQQJJ/p1538194789000100

**Comment by hmstepanek** [comment]
An investigation needs to be done into how we can best address this problem without completely overhauling the current implementation.

**Comment by hmstepanek** [comment]
It should come as no surprise that we can massively reduce the rendering time by not including the content of each timeframe bucket in the template but rather send a request to get that info when the user clicks on it. This is the way elasticsearch is intended to be used in combination with a db anyway.  You get the ids from elasticsearch, then when the user wants to see the details, you run a db query to get the metadata. I don't think this would be too complicated to add; just some simple javascript to call the api when the user expands a bucket. Since there are typically multiple annotations per bucket, I'd recommend adding an api endpoint that allows you to search for multiple ids at once, otherwise we'll be running a bunch of single db queries for each annotation which will be slower.

**Comment by hmstepanek** [comment]
Another pain point I noticed is there are python functions being passed to the jinja template. This is something that Django doesn't let users do at all for probably good reason. I did notice a slight performance improvement when I computed things before passing them to the template but not a lot. I don't think moving them out of the template has specific performance improvement advantages but from the perspective of keeping things simple, and also being able to monitor the performance, I'd recommend moving these out of the template. When they are called from the template that get shoved into rendering time, whereas if they are outside they are represented as application time which means we can better categorize our performance more effectively. They are simply generating url links for tags, users, etc. We could also make them extensions but my thinking here is if we can get them all in one spot (they are all called inside the big loop over the buckets) that we could potentially use some parallelism to get a performance improvement. I think it would also just generally make things more organized/readable if they were in the same spot because as we know, activity pages are disorganized and hard to read. There is giant multi-nested loop in h.activity.query.py:execute that adds all the data/configures the search_results to be passed to the template. This is a pretty expensive loop that could probably be executed in parallel on a per-bucket basis. Note it's also expensive because it queries for all the metadata for each annotation inside the bucket that we may or may not actually need to display. I think if we simply push the grabbing of this data till the user actually requests it, this would have a huge positive performance impact. Not only because we would reduce rendering time in the template but also because this loop would be cheaper possible even entirely unnecessary.

**Comment by robertknight** [comment]
As I think I mentioned before, generating as much HTML as we do on activity pages causes not only slow rendering but also causes downstream issues: Lots of bytes to transfer over the network, lots of HTML for the browser to parse, lots of DOM elements to keep in memory. I would definitely be inclined to focus efforts on resolving that first.

>  I don't think this would be too complicated to add; just some simple javascript to call the api when the user expands a bucket. 

Yes, something like this. Each collapsible row representing a "bucket" of results would need a JS controller, which it already has (`controllers/search-bucket-controller.js`). A minimal implementation would I think involve replacing the rendered annotations with a `data-` attribute or JSON blob including a list of annotation IDs for each bucket, and then requesting rendered annotations when a user expands the bucket and inserting the returned HTML into the appropriate place.

The h website currently doesn't make any requests to the Hypothesis API but uses non-API AJAX endpoints in a few places. While we could use the public API + client-side rendering to do this I think it would be quicker and easier to start with a non-API/non-general purpose endpoint which returns exactly the same HTML as is currently rendered for a bucket of annotations in activity pages.

One other caveat is that activity pages at the moment, in theory, work without JavaScript. You can try this by appending `nojs=1` to any page or turning JS off in your browser. For example: https://hypothes.is/search?nojs=1#. While there are benefits to supporting search result pages that work without JS, I think we should reevaluate if/how that is done.

>  I don't think moving them out of the template has specific performance improvement advantages but from the perspective of keeping things simple, and also being able to monitor the performance, I'd recommend moving these out of the template. When they are called from the template that get shoved into rendering time, whereas if they are outside they are represented as application time which means we can better categorize our performance more effectively. 

Making it easier to measure where time is spent seems like a good reason to try to compute data before template rendering. The earlier point about performance of passing only data to the templates vs triggering computation from within the templates would need more data before making substantial changes. It isn't immediately obvious to me why there would be a significant performance difference.

**Comment by hmstepanek** [comment]
> It isn't immediately obvious to me why there would be a significant performance difference.

To that last point there isn't a significant performance improvement it's more a readability, best practices, and reporting improvement. It's possible the sight perf improvement I noticed was completely negligible. 

I gave this a try locally (removing the presenter calls to fill out the annotation details in query execute and removing the building of the html for the content of each bucket) and was able to get a 3x performance improvement. I'm happy starting with this performance improvement as a first step and seeing where that gets us. There are some more lower impact things I've identified on top of this (I've filed separate issues for them) that I think will help us get the remainder of the way there (fixing fetch annotations, readable groups, pagination, etc).
